### PR TITLE
refactor: Default TCP request pool size

### PR DESCRIPTION
**Pull request by kthui**
## Overview:

Dynamically size the fixed TCP worker pool from the local engine’s effective sequence capacity.

When no explicit override is configured, the worker reports its normalized `engine_max_num_seqs` during registration and the TCP server starts with `ceil(1.5 × engine_max_num_seqs)` workers. This keeps TCP admission aligned with backend capacity while preserving the existing `10,000` fallback.

## Details:

TCP worker-pool sizing now follows this precedence:

1. `DYN_ENGINE_REQUEST_LIMIT`
2. `DYN_TCP_WORKER_POOL_SIZE`
3. `ceil(1.5 × engine_max_num_seqs)`
4. `10,000` fallback

The default `DYN_TCP_WORK_QUEUE_SIZE` is changed to `16`.

Additional changes:

- Add `engine_max_num_seqs` as process-local registration metadata, separate from the existing backend-dependent `max_num_seqs`.
- Normalize engine capacity in Python vLLM, Python SGLang, TensorRT-LLM, native vLLM, SGLang sidecar, and mocker integrations.
- Ensure capacity-bearing registration completes before the shared TCP server is initialized.
- Keep the TCP worker pool fixed after startup; this change does not introduce runtime resizing.
- Preserve existing Python and PyO3 positional-constructor compatibility.
- Add backend-focused tests proving each integration supplies the expected normalized capacity.
- Add a process-isolated mocker E2E covering automatic sizing and both environment-variable overrides.

Validation included:

- Automatic sizing and override-precedence mocker E2E
- Backend capacity unit tests
- Rust formatting and Clippy with warnings denied
- Applicable pre-commit hooks
- Real GPU-backed vLLM worker: `max_num_seqs=4` produced TCP pool capacity `6`
- Real GPU-backed TensorRT-LLM worker: `max_batch_size=4` produced TCP pool capacity `6`
- Both real workers initialized with TCP queue size `16` and successfully registered their endpoints

## Where should the reviewer start?

1. `lib/runtime/src/pipeline/network/ingress/shared_tcp_endpoint.rs`
   - Pool-size precedence, automatic `1.5×` sizing, and queue defaults.

2. `lib/llm/src/local_model.rs`
   - Passes normalized engine capacity into the TCP server before endpoint serving begins.

3. `lib/llm/src/local_model/runtime_config.rs`
   - Defines and consumes `engine_max_num_seqs`.

4. Backend capacity adapters:
   - `components/src/dynamo/vllm/capacity.py`
   - `components/src/dynamo/sglang/capacity.py`
   - `components/src/dynamo/trtllm/utils/trtllm_utils.py`
   - `lib/vllm-rs-backend/src/backend.rs`
   - `lib/sglang-sidecar/src/engine.rs`

5. `tests/router/test_tcp_worker_pool_sizing_e2e.py`
   - Process-level automatic-sizing and override-precedence coverage.

## Related Issues

N/A

### CODE REVIEW

File: tests/router/test_router_e2e_with_mockers.py [hunk 1]
```diff
@@ -947,6 +973,279 @@ def test_router_decisions_disagg(
...
+        with FrontendRouterProcess(
+            request=request,
+            block_size=1,
+            frontend_port=frontend_port,
+            namespace=shared_namespace,
+            enforce_disagg=True,
+            request_plane=request_plane,
+            min_initial_workers=1,
+        ):
+            asyncio.run(exercise_gate())
```
[the remark this answers is not in the document]
Comment (kthui) [reply tests/router/test_router_e2e_with_mockers.py]: Addressed in [`572b94f`](https://github.com/ai-dynamo/dynamo/commit/572b94fdf6168c78686ab84a61f39003217de22a). The reusable Prometheus parser now lives in `tests/router/helper.py`, and the core four-request prefill-admission/decode-gating scenario now lives in `tests/router/common.py`. `test_router_e2e_with_mockers.py` retains the SGLang mocker configuration and process wrapper and delegates the scenario assertions to the shared helper. The focused `[2-tcp]` E2E passed after the extraction (`1 passed in 19.91s`), along with Black, Python compilation, and `git diff --check`.

File: tests/router/test_router_e2e_with_mockers.py [hunk 2]
```diff
@@ -947,6 +973,279 @@ def test_router_decisions_disagg(
...
+        prefill_mocker_args=prefill_mocker_args,
+        decode_mocker_args=decode_mocker_args,
+        num_prefill_mockers=1,
+        num_decode_mockers=1,
+        enable_disagg_bootstrap=True,
+        request_plane=request_plane,
+        event_plane="nats",
+        prefill_env_overrides=prefill_env,
+        decode_env_overrides=decode_env,
+    ) as (prefill_workers, decode_workers):
```
[the remark this answers is not in the document]
Comment (kthui) [reply tests/router/test_router_e2e_with_mockers.py]: Addressed in [`572b94f`](https://github.com/ai-dynamo/dynamo/commit/572b94fdf6168c78686ab84a61f39003217de22a): the context-manager targets are now `_prefill_workers` and `_decode_workers`. Black, Python compilation, the focused `[2-tcp]` E2E, and `git diff --check` all passed.

File: tests/router/test_router_e2e_with_mockers.py [hunk 3]
```diff
@@ -947,6 +973,280 @@ def test_router_decisions_disagg(
...
+                    frontend_metrics_url,
+                    "dynamo_request_plane_send_seconds_count",
+                    lambda value: value >= frontend_ack_baseline + 1,
+                )
+                if dispatcher_task.done():
+                    raise AssertionError(
+                        "dispatcher-held request unexpectedly established a response stream"
+                    )
+
+                queued_task = asyncio.create_task(
```
[the remark this answers is not in the document]
Comment (kthui) [reply tests/router/test_router_e2e_with_mockers.py]: Addressed in [`991b8ed`](https://github.com/ai-dynamo/dynamo/commit/991b8ed58941ebd335190e657e0c44ae5686d404). The scenario now waits for `dynamo_work_handler_queue_depth == 0` after the first filler ACK and before submitting `queued_task`, which establishes that the first filler is dispatcher-held before the second filler occupies the bounded queue. The scenario had moved to `tests/router/common.py` in the preceding refactor.

The focused `[2-tcp]` E2E passed three consecutive runs with the barrier (`20.71s`, `20.21s`, and `20.00s`), along with Black, Python compilation, and `git diff --check`.

I also applied the barrier-hardened test to detached `e427241` (the parent of production gate fix `39d2a6899609`) as a negative control. It failed twice at the intended no-decode oracle (`19.98s` and `19.90s`), both with `status=None, send_before=5.0, send_after=7.0`, confirming the deterministic staging still catches the pre-fix extra decode dispatch.

**Comment by kthui** [comment]
Closing this PR as the scope has changed. See the new PR #11851
